### ISSUE TITLE
D-048 — V0.5 extended-exposure shadow consequence-learning curve

### ISSUE THREAD

**Issue by PiFlow** [PiFlow/aweform#170]
# D-048 — V0.5 extended-exposure shadow consequence-learning curve

**Status:** AUTHORIZED DEVELOPMENT IMPLEMENTATION — D-048 only  
**Lane:** Development / shadow-learning diagnostic  
**Authorized base:** `c401aaa0c6a60bf3cc13840c40951957854ef2ef`  
**Predecessor:** D-047, merged by PR #169  
**Flow decision:** before adding memory, recurrence, model capacity, confidence machinery, need, utility, or behavioral influence, test whether the unchanged D-046 learner is simply still learning when given substantially more continuous experience.

## Scientific question

If Aweform is given **more experience but no additional learner capability**, does the existing D-046 consequence learner continue to improve, plateau, or deteriorate?

D-048 asks for the learning curve of the unchanged V0.5 shadow consequence learner across cumulative exposure checkpoints:

- `0×` = 0 learning transitions;
- `1×` = 564 transitions;
- `2×` = 1,128 transitions;
- `4×` = 2,256 transitions;
- `8×` = 4,512 transitions.

The central question is not whether a larger or more sophisticated model could perform better. It is whether the **current** 528-weight D-046 learner has reached its useful learning limit under the existing experience distribution.

D-048 changes **exposure duration**, not learner capability.

## Why this is the next question

D-047 found mixed but real action-consequence discrimination after the original D-046 exposure:

- wheel-delta non-tie direction agreement was about 98.6%;
- energy was about 81.4%;
- beacon channels were about 56.6–63.3%;
- temperature/contact had sparse non-tie support and weak exact-tie behavior.

Those results do not tell us whether the learner architecture is inadequate or whether D-046 simply stopped training too early.

Before adding history, recurrence, a larger predictor, a different representation, confidence gating, utility, need, or behavioral use, D-048 measures how the unchanged learner changes with more experience.

## Lifecycle and initialization

D-048 is a new Development stage and therefore a **new lifetime** under ADR 0010.

For every D-048 seed:

- initialize a fresh full D-046 consequence learner from exact zero weights;
- initialize a fresh matched D-046 state-only learner from exact zero weights;
- initialize an unchanged D-045 environment from the same D-046 initial physical state;
- do **not** import or continue D-046/D-047 learned weights across the D-stage boundary.

Within each D-048 lifetime, however, the organism/environment/learner are continuous for the entire 4,512-transition exposure. There is no learner reset, environment reset, RNG reseed, hidden stage signal, or fresh-start marker between exposure passes or checkpoints.

Evaluator-only checkpoint evaluation must not alter that continuous causal state.

## Frozen learner and organism boundary

Reuse the D-045/D-046 semantics exactly.

D-048 may not change:

- D-045 physics, embodiment, contact, energy/thermal bookkeeping, reward, `info`, observations, or action semantics;
- the exact eight organism-visible channels;
- the two raw-wheel command inputs;
- the D-046 model-state normalization;
- the 66-feature quadratic feature map;
- the 528 full-learner weights;
- zero initialization;
- D-046 NLMS step `0.5`;
- D-046 update timing;
- D-046 state-only comparator semantics;
- the D-046 raw-wheel action envelope;
- any organism-visible information boundary.

Predictions remain strictly **shadow-only**. They may not influence command choice, termination, reward, observations, viability, or any behavioral arbitration.

## Development seeds

Use exactly the existing legal D-046 development seeds:

`21046..21065`

No new Development seeds and no formal reserved seeds are required.

Validate the seed block against all current formal reservations.

These are Development results only. Reuse of these seeds is not confirmatory evidence.

## Exposure protocol

For each seed, build the exact existing 564-transition D-046 curriculum for that seed.

Use that same ordered 564-command sequence as one **exposure pass**.

Run exactly **8 consecutive passes** in one continuous D-045 lifetime:

`8 × 564 = 4,512` learning transitions per seed.

Important:

- do not reset the environment between passes;
- do not reset either learner between passes;
- do not reshuffle or tune the curriculum after seeing output;
- do not change the curriculum contents, command amplitudes, block definitions, prefix/suffix, or seed-derived order;
- repeating the established curriculum is deliberate: D-048 isolates **amount of experience under the same experience distribution** before testing broader coverage.

If the unchanged D-045 lifetime terminates, truncates, becomes nonviable, or otherwise cannot legally reach a planned checkpoint, **do not repair this by increasing battery, changing physics, resetting the environment, or skipping transitions**. Preserve the termination and the last legally reached checkpoint as the D-048 result for that seed/lifetime.

## Checkpoints

Evaluate at exactly:

`0×, 1×, 2×, 4×, 8×`.

The `0×` checkpoint occurs after initialization and before any learner update.

At each checkpoint:

1. snapshot the full and state-only learned state read-only;
2. record weight digests and bounded summaries;
3. evaluate snapshots in isolated evaluator branches;
4. do not update the live learner from checkpoint evaluation;
5. prove that the live environment, learner weights, retained state, and future exposure trajectory are unchanged by evaluation;
6. resume the same continuous lifetime.

Checkpoint evaluation is evaluator-only and must be causally invisible to the learner.

## Held-out evaluation

At every checkpoint use the exact D-046/D-047 held-out matrix only:

- poses `H0..H8`;
- raw-wheel actions `A0..A8`.

Do not add, delete, tune, or replace held-out poses/actions after seeing D-048 output.

For each seed/pose, all nine candidate actions must start from the same isolated D-045 physical state. Real candidate consequences are evaluated in independent one-step branches exactly as in D-047. No held-out candidate consequence becomes learning experience.

### A. Action-consequence discrimination

For the full learner, reproduce the D-047 pairwise contrast analysis at each checkpoint:

`predicted_contrast = predicted_delta(A_i) - predicted_delta(A_j)`

`actual_contrast = actual_delta(A_i) - actual_delta(A_j)`

with fixed ascending action-ID orientation.

Retain per channel:

- contrast MAE;
- exact negative/zero/positive sign agreement;
- actual tie count;
- predicted tie count;
- non-tie sign-agreement rate when defined;
- support.

At minimum retain pooled results and sufficient compact strata to determine whether learning changes are general or concentrated by:

- seed;
- held-out pose;
- unordered action pair;
- boundary involvement;
- contact-transition involvement.

The state-only and zero-change action-contrast comparators remain structurally action-indifferent and therefore zero contrast from the same state.

### B. Absolute one-step prediction

At each checkpoint also retain D-046-style held-out one-step prediction metrics separately for:

- full learner;
- matched state-only learner;
- zero-change comparator.

For all eight channels retain:

- one-step MAE;
- learned-minus-zero-change MAE;
- charging-contact Brier score where applicable;
- support.

This is required because directional discrimination can improve while magnitude calibration does not.

## Learning-progress diagnostics

The goal is to see **where the learner is on its learning curve**, not merely compare 1× with 8×.

For each of the 8 exposure passes and each seed, record compact pre-update/prequential training diagnostics using the unchanged D-046 prediction-before-update semantics.

At minimum retain per channel:

- prequential MAE for that pass;
- matched state-only prequential MAE;
- zero-change MAE;
- contact-event/non-event support;
- contact-entry and contact-exit counts;
- charging-positive support;
- boundary-event support if available without changing D-045.

At each checkpoint retain:

- complete full-learner weight digest;
- complete state-only weight digest;
- weight count/finite check;
- L2 norm;
- L2 change from the previous checkpoint for each learner;
- whether checkpoint evaluation changed either digest (must be false).

Do not combine channels into one scalar learning score.

## Internal comparison discipline

The primary D-048 inference is **within the same D-048 execution/runtime across exposure checkpoints**.

Because D-046 previously demonstrated architecture-dependent floating-point digest differences across platforms, do not require D-048 `1×` weights to cryptographically equal the historical D-046 artifact across architectures.

Instead:

- require exact deterministic regeneration of the D-048 artifact from its own frozen executable/runtime;
- require the exact D-046 curriculum sequence semantics and learner/update semantics;
- report D-047 values only as historical context, not as an exact numeric identity gate for the D-048 `1×` checkpoint.

## Interpretation boundary

There is **no universal PASS threshold** and no selected “best checkpoint.”

All checkpoints remain part of the result.

Report each channel separately and preserve the families:

- interoceptive: energy, temperature;
- exteroceptive: beacon left, beacon forward, beacon right, charging contact;
- proprioceptive: wheel delta left, wheel delta right.

For each channel describe the observed curve as appropriate, e.g. continuing improvement, approximate plateau, deterioration/interference, sparse/indeterminate support, or mixed. Such descriptions must follow the recorded metrics and must not be converted into a scalar competence score.

Especially:

- stronger wheel learning does not establish exteroceptive or viability readiness;
- poor contact results with sparse contact-transition exposure do not establish model-capacity failure;
- a plateau does not by itself establish that recurrence/history or a larger learner is required;
- continued improvement does not authorize prediction-driven behavior.

D-048 itself must not choose a successor.

## Required causal controls

At minimum verify:

- exact authorized base ancestry;
- exact D-046 curriculum definitions reused unchanged;
- exactly 20 legal development seeds;
- fresh zero initialization for both learners at D-048 lifetime start;
- exactly 8 continuous 564-transition passes unless a genuine unchanged-D-045 termination prevents completion;
- no environment/learner reset between passes or checkpoints;
- checkpoint evaluation leaves the live learner/environment causal state unchanged;
- held-out evaluation never updates weights;
- held-out observations/outcomes never enter later training;
- learner updates use only closure-valid D-046 organism-visible state/action/consequence information;
- reward remains exactly `0.0`;
- organism-facing `info == {}`;
- no evaluator seed label, pose, geometry, branch result, checkpoint label, exposure multiplier, future consequence, or metric reaches model input;
- historical D-001..D-047 source/records/artifacts remain unchanged.

Where practical, include an identity control showing that a lifetime run with checkpoint instrumentation disabled produces the same final causal trajectory/learner state as the instrumented run.

## Required outputs

Prefer the smallest implementation:

- `src/aweform/d048.py`
- `tests/test_d048.py`
- `development/D-048-v05-extended-exposure-shadow-learning-curve.md`
- compact deterministic JSON artifact with the same stem
- one D-048 row in `development/INDEX.md` only in the result layer after official execution

Do not modify D-045, D-046, or D-047 source unless a genuine blocker is found; stop and return to Flow/Sol rather than silently refactoring predecessor semantics.

Do not create a stage-specific visualizer unless Flow separately requests one.

The artifact should remain compact. It should preserve learning-curve aggregates and provenance, not raw trajectories. Full checkpoint weight vectors are not required; exact digests plus deterministic regeneration are sufficient unless implementation demonstrates a concrete reproducibility need.

## Freeze / execution discipline

Use the corrected D-047 provenance pattern from the start:

1. implement the D-048 runner, tests, schema, controls, and a result-free protocol record;
2. commit that executable-only protocol as a clean commit;
3. **push it and verify the exact 40-character commit is retrievable from GitHub before executing the official run**;
4. run the official D-048 audit from that exact pushed SHA;
5. inspect and summarize the descriptive result without changing protocol;
6. commit only the deterministic artifact, completed D-record/result layer, and D-048 ledger row afterward;
7. independently regenerate the artifact from the same frozen executable SHA and require byte-identical equality.

If a protocol/artifact defect is found after execution, preserve invalidated provenance, correct without result-driven tuning, freeze/push a new executable SHA, and rerun from scratch.

## Validation before handoff

Require:

- focused D-048 tests;
- full `pytest -q`;
- `ruff check .`;
- strict `mypy src --strict`;
- Python compile/import checks;
- `git diff --check`;
- authorized-base ancestry;
- exact executable-freeze GitHub retrievability;
- exact curriculum/pass/checkpoint counts;
- seed-reservation validation;
- checkpoint causal-state identity/read-only controls;
- instrumented-versus-uninstrumented identity control where practical;
- deterministic artifact byte-identical regeneration;
- exact-current-HEAD GitHub CI/checks.

## Explicit non-goals / STOP boundaries

D-048 does **not** authorize:

- changing D-045 physics, battery, temperature, dock, sensors, or action semantics;
- changing D-046 learner architecture, feature map, weight count, learning rate, initialization, or update rule;
- importing D-046/D-047 learned weights into the D-048 lifetime;
- broader/new calibration actions or experience distribution;
- curriculum adaptation based on observed performance;
- history or recurrence;
- larger/deeper predictors;
- new sensors or organism-visible channels;
- confidence gating;
- prediction-driven action selection;
- choosing/ranking a “best” action;
- utility, value, desirability, need satisfaction, homeostatic/allostatic objective;
- planner, search, MPC, tree rollout, RL, Q-learning, policy gradients;
- intrinsic reward, curiosity, or play;
- replay buffers or offline retraining;
- changing behavior based on checkpoint metrics;
- EXP execution or confirmatory claims;
- D-049 or any successor.

If any such change appears necessary, STOP and return to Flow/Sol.

## Governance

D-048 is ordinary Development-lane work inside the accepted V0.5 and ADR-0010 lifetime-plasticity boundaries. It changes exposure duration for an already accepted shadow learner but does not change a durable organism architecture, sensory/plasticity permission, information boundary, safety boundary, frozen evidence, or reserved-seed contract.

Therefore the formal ADR-0013 two-independent-reviewer gate is not currently required if implementation stays exactly within this issue. Normal exact-current-HEAD independent Sol review applies. Flow alone authorizes merge.

Repository evidence and exact issue/PR state outrank stale current-state prose elsewhere in the repository. Do not bundle unrelated documentation cleanup into D-048.

---

## Luna Manager authorization

```text
REPOSITORY: PiFlow/aweform
TASK: D-048
ISSUE: THIS_ISSUE
BASE_HEAD: c401aaa0c6a60bf3cc13840c40951957854ef2ef
EXPECTED_BRANCH: codex/d048-v05-extended-exposure-shadow-learning-curve
EXPECTED_EXISTING_PR: NONE
FRESH_WORKER_MECHANISM_EDIT: codex exec --ephemeral -m gpt-5.6-luna -s workspace-write

SOL_TASK_BLOCK_BEGIN
Implement D-048 exactly as authorized by this issue from BASE_HEAD c401aaa0c6a60bf3cc13840c40951957854ef2ef.

Create a fresh D-048 lifetime for each existing development seed 21046..21065 with the unchanged D-045 substrate and fresh zero-initialized D-046 full + state-only learners. Repeat the exact seed-specific D-046 564-transition curriculum for 8 consecutive passes without resetting environment or learner, giving 4,512 continuous learning transitions per completed seed.

Evaluate read-only snapshots at exactly 0×, 1×, 2×, 4×, and 8× exposure. At every checkpoint run the exact D-046/D-047 H0..H8 × A0..A8 held-out matrix without feeding evaluation into learning. Retain D-047 action-pair discrimination metrics and D-046-style absolute one-step metrics, plus per-pass prequential learning diagnostics and compact weight-change/digest diagnostics.

The scientific output is the channel-specific learning curve: whether the unchanged learner continues improving, plateaus, deteriorates, or remains indeterminate under sparse support. There is no universal pass threshold and no best checkpoint.

Keep predictions entirely shadow-only. Do not change learner architecture, feature map, learning rate, sensors, actions, physics, curriculum distribution, history, recurrence, capacity, confidence, utility, need, action selection, planning, reward, EXP work, or any successor.

Use a two-layer provenance history: push and verify a clean executable-only freeze before the official run, then commit only artifact/completed record/index in the result layer. Require byte-identical regeneration from the frozen pushed SHA.

Open exactly one PR and post a top-level [LUNA-HANDOFF] naming the exact current 40-character PR HEAD. Do not merge. Do not start D-049.
SOL_TASK_BLOCK_END

GOVERNANCE_SUMMARY_BEGIN
Ordinary Development-lane extended-exposure diagnostic inside accepted V0.5 and ADR-0010 boundaries. Exposure duration changes; learner capability and organism/information/safety boundaries do not. No formal ADR-0013 two-reviewer gate unless implementation discovers a durable boundary change. Normal exact-current-HEAD independent Sol review applies. Flow controls merge. No successor is authorized.
GOVERNANCE_SUMMARY_END

MANAGER_RULES_BEGIN
Before launching the fresh worker, verify GitHub main still equals BASE_HEAD exactly, the expected branch does not already contain work, and no competing D-048 PR exists. If main moved before implementation begins, STOP and request renewed authorization rather than silently rebasing. Use one fresh Luna worker, one branch, one PR. Freeze/push the executable-only commit before official execution. Do not merge. Do not start D-049.
MANAGER_RULES_END
```
